torchmetrics/metric.py

```diff
@@ -515,6 +515,10 @@ def _apply(self, fn: Callable) -> Module:
                     "Expected metric state to be either a Tensor" f"or a list of Tensor, but encountered {current_val}"
                 )
 
+        # make sure to update the device attribute
+        # if the dummy tensor moves device by fn function we should also update the attribute
+        self._device = fn(torch.zeros(1, device=self.device)).device
```

Comment:
Why is this needed at all. Shouldn't this be handled by our wrapped `.to` function?

Reply:
Because if you call `.to` on a parent module, it does actually not call the `.to` method of the child modules but it instead call `_apply` method. This change actually makes the wrapped `cuda`, `cpu`, `to` we have right now obsolete.

Reply:
I see, thanks :) But then do we also want to get the dtype there or is that something we only want with an explicit call of `.to` etc?

Reply:
the `dtype` casting was actually remove in a earlier PR, because users were worried that if they run mixed-precision training that their metrics would be automatically casted to half precision given them wrong results. It can still be done but you need to explicitly call the special method `set_type`.

Reply:
Should we then extend this PR to remove the legacy wrappings for the device changes? Since this won't change behaviour or API it should be fine, right?

Reply:
@justusschock you are right, it is removed now :]